torch/_inductor/simple_fsdp/greedy_bucket.py

```diff
@@ -0,0 +1,379 @@
+            enable_bucket = forward_greedy(current_comp, current_ag, current_mem, 
+                node_dep_dict[node]["COMPUTE_TIME"], node_dep_dict[node]["AG_TIME"], 
+                node_dep_dict[node]["MEMORY"], 160, first_AG)
+            first_AG = False
+            if enable_bucket:
+                # merge all gather
+                merged_all_gather, ag_buffer = merge_allgather(sched, all_gather_list)
+                merged_wait = merge_ag_wait(sched, ag_wait_list, all_gather_list, ag_buffer)
+                for n in all_gather_dep_list + [merged_all_gather, merged_wait]:
+                    if n not in result_list:
+                        result_list.append(n)
+                compute_list = [i for i in compute_list if i not in result_list]
+                result_list.extend(compute_list)
+
+                # clear the list for bucketing
+                all_gather_list = [node]
+                all_gather_dep_list = node_dep_dict[node]["AG_DEP"]
+                ag_wait_list = node_dep_dict[node]["AG_WAIT"]
+                compute_list = node_dep_dict[node]["COMPUTE"]
+
+                # clear the number for greedy
+                current_comp = next_comp
+                current_mem = node_dep_dict[node]["MEMORY"]
+                current_ag = node_dep_dict[node]["AG_TIME"] 
+                next_comp = node_dep_dict[node]["COMPUTE_TIME"]
+            else:
+                # update the list for bucketing
+                all_gather_list.append(node)
+                all_gather_dep_list.extend(node_dep_dict[node]["AG_DEP"])
+                ag_wait_list.extend(node_dep_dict[node]["AG_WAIT"])
+                compute_list.extend(node_dep_dict[node]["COMPUTE"])
+
+                # update the number for greedy
+                current_mem += node_dep_dict[node]["MEMORY"]
+                current_ag += node_dep_dict[node]["AG_TIME"]
+                next_comp += node_dep_dict[node]["COMPUTE_TIME"]
+    
+    if len(all_gather_list) > 0:
+        merged_all_gather, ag_buffer = merge_allgather(sched, all_gather_list)
+        merged_wait = merge_ag_wait(sched, ag_wait_list, all_gather_list, ag_buffer)
+        for n in all_gather_dep_list + [merged_all_gather, merged_wait]:
+            if n not in result_list:
+                result_list.append(n)
+    compute_list = [i for i in compute_list if i not in result_list]
+    result_list.extend(compute_list)
+    return result_list
+
+
+def forward_greedy(current_comp, current_ag, current_mem, node_comp, node_ag, node_mem, memory_constraint, first_AG):
+    """
+    Greedy algorithm for forward
+    """
+    enable_bucket = False
+
+    if first_AG:
+        enable_bucket = True
+
+    if current_mem + node_mem > memory_constraint:
+        print("memory constraint not satisfied")
+        enable_bucket = True
+
+    if current_ag + node_ag > current_comp:
+        print("time constraint not satisfied")
+        enable_bucket = True
+    
+    if current_ag == 0:
```

Comment:
This happens when we go through the first few compute nodes, which doesn't belong to any AG node. 

The current_ag is 0. We cannot bucket them, because there are no AG/RS to bucket.

We can do this when checking if this is the first AG, a.k.a., change this to `first_AG and current_ag != 0`